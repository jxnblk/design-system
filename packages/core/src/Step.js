import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Button from './Button'
import Text from './Text'
import Icon from './Icon'
import { getPaletteColor } from './utils'

const StepperFlex = styled(Button)`
  display: flex;
  background-color: ${getPaletteColor('background.lightest')};
  color: ${getPaletteColor('base')};
  cursor: ${props => (props.onClick ? 'pointer' : 'default')};
  padding: 0;
  font-weight: unset;

  &:hover {
    background-color: ${getPaletteColor('background.lightest')};
  }
`

function Step({ className, active, completed, children, onClick }) {
  const color = active || completed ? 'primary' : 'text.light'

  return (
    <StepperFlex
      className={className}
      disabled={!onClick}
      alignItems="center"
      onClick={onClick}
    >
      {completed && (
        <Icon
          name="Check"
          color="primary"
          size={16}
          mr={1}
          aria-hidden={false}
          aria-label="Step Completed"
        />
      )}
      <Text color={color} bold={active} fontSize={1} aria-selected={active}>
        {children}
      </Text>
    </StepperFlex>
  )
}

Step.displayName = 'Step'

Step.propTypes = {
  className: PropTypes.string,
  active: PropTypes.bool,
  completed: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func
}

Step.defaultProps = {
  className: '',
  active: false,
  completed: false
}

export default Step
