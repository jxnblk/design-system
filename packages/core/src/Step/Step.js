import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { Check } from 'pcln-icons'

import { Button } from '../Button'
import { Text } from '../Text'
import { getPaletteColor } from '../utils'

const StepperFlex = styled(Button)`
  display: flex;
  background-color: ${getPaletteColor('background.lightest')};
  color: ${getPaletteColor('base')};
  cursor: ${(props) => (props.onClick ? 'pointer' : 'default')};
  padding: 0;
  font-weight: unset;

  &:hover {
    background-color: ${getPaletteColor('background.lightest')};
  }
`

function Step({ className, active, completed, children, onClick, ...props }) {
  const color = active || completed ? 'primary' : 'text.light'

  return (
    <StepperFlex className={className} alignItems='center' onClick={onClick} {...props}>
      {completed && (
        <Check color='primary' size={16} mr={1} aria-hidden={false} aria-label='Step Completed' />
      )}
      <Text color={color} bold={active} fontSize={1} aria-selected={active}>
        {children}
      </Text>
    </StepperFlex>
  )
}

Step.displayName = 'Step'

Step.propTypes = {
  className: PropTypes.string,
  active: PropTypes.bool,
  completed: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
}

Step.defaultProps = {
  className: '',
  active: false,
  completed: false,
}

export default Step
