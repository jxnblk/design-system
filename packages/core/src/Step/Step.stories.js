import React from 'react'
import { storiesOf } from '@storybook/react'
import { withInfo } from '@storybook/addon-info'
import { action } from '@storybook/addon-actions'

import { Step } from '.'

const onClick = action('Step Clicked')

storiesOf('Step', module)
  .add(
    'Step component',
    withInfo({
      inline: true,
      text: 'Use the <Stepper.Step /> component to render a step.',
    })(() => <Step>Step</Step>)
  )
  .add('Current Step', () => <Step active>Step</Step>)
  .add('Completed Step', () => <Step completed>Step</Step>)
  .add('Current and Completed Step', () => (
    <Step active completed>
      Step
    </Step>
  ))
  .add('Clickable Step', () => <Step onClick={onClick}>Step</Step>)
