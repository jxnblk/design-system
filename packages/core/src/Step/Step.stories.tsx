import React from 'react'
import { action } from '@storybook/addon-actions'

import { Step } from '.'

const onClick = action('Step Clicked')

export default {
  title: 'Step',
  component: Step,
  parameters: {
    docs: {
      description: {
        component: 'Use the <Stepper.Step /> component to render a step.',
      },
    },
  },
}

export const StepComponent = () => <Step>Step</Step>

export const CurrentStep = () => <Step active>Step</Step>
export const CompletedStep = () => <Step completed>Step</Step>

export const CurrentAndCompletedStep = () => (
  <Step active completed>
    Step
  </Step>
)

export const ClickableStep = () => <Step onClick={onClick}>Step</Step>
