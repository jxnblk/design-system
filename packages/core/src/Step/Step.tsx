import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { Check } from 'pcln-icons'

import { Button } from '../Button'
import { Text } from '../Text'
import { getPaletteColor } from '../utils'

const StepButton = styled(Button)`
  display: flex;
  background-color: ${getPaletteColor('background.lightest')};
  color: ${getPaletteColor('base')};
  cursor: ${(props) => (props.onClick ? 'pointer' : 'default')};
  padding: 0;
  font-weight: unset;

  &:hover {
    background-color: ${getPaletteColor('background.lightest')};
  }
`

const propTypes = {
  className: PropTypes.string,
  active: PropTypes.bool,
  completed: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
}

export interface IStepProps {
  className?: string
  active?: boolean
  completed?: boolean
  children?: React.ReactNode | string
  onClick?: (unknown) => unknown
}

const Step: React.FC<IStepProps> = ({ className, active, completed, children, onClick, ...props }) => {
  const color = active || completed ? 'primary' : 'text.light'

  return (
    <StepButton className={className} alignItems='center' onClick={onClick} mr={4} {...props}>
      {completed && (
        <Check color='primary' size={16} mr={1} aria-hidden={false} aria-label='Step Completed' />
      )}
      <Text color={color} bold={active} fontSize={1} aria-selected={active}>
        {children}
      </Text>
    </StepButton>
  )
}

Step.displayName = 'Step'

Step.propTypes = propTypes

Step.defaultProps = {
  className: '',
  active: false,
  completed: false,
}

export default Step
