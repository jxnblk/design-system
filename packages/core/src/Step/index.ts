export { default as Step, IStepProps } from './Step'
