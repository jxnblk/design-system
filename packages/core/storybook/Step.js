import React from 'react'
import { storiesOf } from '@storybook/react'
import { withInfo } from '@storybook/addon-info'
import Step from '../src/Step'

const onClick = () => alert('Step Clicked')

storiesOf('Step', module)
  .add(
    'Step component',
    withInfo({
      inline: true,
      text: 'Use the <Step /> component to render a step.'
    })(() => <Step>Step</Step>)
  )
  .add('Current Step', () => <Step isCurrent>Step</Step>)
  .add('Completed Step', () => <Step isCompleted>Step</Step>)
  .add('Current and Completed Step', () => (
    <Step isCurrent isCompleted>
      Step
    </Step>
  ))
  .add('Clickable Step', () => <Step onClick={onClick}>Step</Step>)
