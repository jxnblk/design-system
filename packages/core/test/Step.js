import React from 'react'
import Step from '../src/Step'

const hasIcon = component => {
  const { container } = renderWithTheme(component)
  return container.querySelector('svg')
}

describe('Step', () => {
  test('renders without crashing', () => {
    const json = rendererCreateWithTheme(<Step>Step</Step>).toJSON()
    expect(json).toMatchSnapshot()
  })

  test('renders active step', () => {
    const component = <Step active>Step</Step>
    const json = rendererCreateWithTheme(component).toJSON()

    expect(json).toMatchSnapshot()
    expect(json).toHaveStyleRule('cursor', 'default')

    expect(hasIcon(component)).toBeFalsy()
  })

  test('renders completed step', () => {
    const component = <Step completed>Step</Step>
    const json = rendererCreateWithTheme(component).toJSON()

    expect(json).toMatchSnapshot()
    expect(json).toHaveStyleRule('cursor', 'default')

    expect(hasIcon(component)).toBeTruthy()
  })

  test('renders as active and completed', () => {
    const component = (
      <Step active completed>
        Step
      </Step>
    )
    const json = rendererCreateWithTheme(component).toJSON()

    expect(json).toMatchSnapshot()
    expect(json).toHaveStyleRule('cursor', 'default')

    expect(hasIcon(component)).toBeTruthy()
  })

  test('renders with on click', () => {
    const component = <Step onClick={() => {}}>Step</Step>
    const json = rendererCreateWithTheme(component).toJSON()

    expect(json).toMatchSnapshot()
    expect(json).toHaveStyleRule('cursor', 'pointer')

    expect(hasIcon(component)).toBeFalsy()
  })
})
