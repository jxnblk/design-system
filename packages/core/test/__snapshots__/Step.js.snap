// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step renders active step 1`] = `
.c1 {
  color: #007aff;
  font-size: 14px;
  font-weight: 700;
}

.c0 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 600;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 2px;
  border-width: 0;
  border-style: solid;
  font-size: 14px;
  padding: 9.5px 18px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  cursor: default;
  padding: 0;
  font-weight: unset;
}

.c0:disabled {
  cursor: not-allowed;
  color: #4f6f8f;
  background-color: #edf0f3;
}

.c0:hover {
  background-color: #fff;
}

<button
  className="c0"
>
  <div
    aria-selected={true}
    className="c1"
    color="primary"
    fontSize={1}
  >
    Step
  </div>
</button>
`;

exports[`Step renders as active and completed 1`] = `
.c3 {
  color: #007aff;
  font-size: 14px;
  font-weight: 700;
}

.c1 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  margin-right: 4px;
  color: #007aff;
}

.c2 {
  outline: none;
}

.c0 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 600;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 2px;
  border-width: 0;
  border-style: solid;
  font-size: 14px;
  padding: 9.5px 18px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  cursor: default;
  padding: 0;
  font-weight: unset;
}

.c0:disabled {
  cursor: not-allowed;
  color: #4f6f8f;
  background-color: #edf0f3;
}

.c0:hover {
  background-color: #fff;
}

<button
  className="c0"
>
  <svg
    aria-hidden={false}
    aria-label="Step Completed"
    className="c1 c2"
    color="primary"
    fill="currentcolor"
    focusable={false}
    height={16}
    role="img"
    tabIndex={-1}
    viewBox="0 0 24 24"
    width={16}
  >
    <path
      d="M8.6 15.6l-4.2-4.2L3 12.8l5.6 5.6 12-12L19.2 5 8.6 15.6z"
    />
  </svg>
  <div
    aria-selected={true}
    className="c3"
    color="primary"
    fontSize={1}
  >
    Step
  </div>
</button>
`;

exports[`Step renders completed step 1`] = `
.c3 {
  color: #007aff;
  font-size: 14px;
}

.c1 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  margin-right: 4px;
  color: #007aff;
}

.c2 {
  outline: none;
}

.c0 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 600;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 2px;
  border-width: 0;
  border-style: solid;
  font-size: 14px;
  padding: 9.5px 18px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  cursor: default;
  padding: 0;
  font-weight: unset;
}

.c0:disabled {
  cursor: not-allowed;
  color: #4f6f8f;
  background-color: #edf0f3;
}

.c0:hover {
  background-color: #fff;
}

<button
  className="c0"
>
  <svg
    aria-hidden={false}
    aria-label="Step Completed"
    className="c1 c2"
    color="primary"
    fill="currentcolor"
    focusable={false}
    height={16}
    role="img"
    tabIndex={-1}
    viewBox="0 0 24 24"
    width={16}
  >
    <path
      d="M8.6 15.6l-4.2-4.2L3 12.8l5.6 5.6 12-12L19.2 5 8.6 15.6z"
    />
  </svg>
  <div
    aria-selected={false}
    className="c3"
    color="primary"
    fontSize={1}
  >
    Step
  </div>
</button>
`;

exports[`Step renders with on click 1`] = `
.c1 {
  color: #4f6f8f;
  font-size: 14px;
}

.c0 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 600;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 2px;
  border-width: 0;
  border-style: solid;
  font-size: 14px;
  padding: 9.5px 18px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  cursor: pointer;
  padding: 0;
  font-weight: unset;
}

.c0:disabled {
  cursor: not-allowed;
  color: #4f6f8f;
  background-color: #edf0f3;
}

.c0:hover {
  background-color: #fff;
}

<button
  className="c0"
  onClick={[Function]}
>
  <div
    aria-selected={false}
    className="c1"
    color="text.light"
    fontSize={1}
  >
    Step
  </div>
</button>
`;

exports[`Step renders without crashing 1`] = `
.c1 {
  color: #4f6f8f;
  font-size: 14px;
}

.c0 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 600;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 2px;
  border-width: 0;
  border-style: solid;
  font-size: 14px;
  padding: 9.5px 18px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  cursor: default;
  padding: 0;
  font-weight: unset;
}

.c0:disabled {
  cursor: not-allowed;
  color: #4f6f8f;
  background-color: #edf0f3;
}

.c0:hover {
  background-color: #fff;
}

<button
  className="c0"
>
  <div
    aria-selected={false}
    className="c1"
    color="text.light"
    fontSize={1}
  >
    Step
  </div>
</button>
`;
